File: net/JetBrains.Serialization/src/AlgorithmInfo.cs
using System;
using System.Collections.Generic;
using JetBrains.Annotations;
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.X509;
using Org.BouncyCastle.Cms;

namespace JetBrains.Serialization;

[JsonObject(MemberSerialization.Fields)]
public class AlgorithmInfo : IEncodableInfo
{
  private String _name;
  [CanBeNull] private IEncodableInfo _additionalValue = null;
  private TextualInfo _algorithmIdentifier;

  public AlgorithmInfo(AlgorithmIdentifier signatureAlgorithm)
  {
    _name = SerializationUtils.AlgorithmNameFromId(signatureAlgorithm.Algorithm);
    _additionalValue = signatureAlgorithm.Parameters?.ToAsn1Object()?.ToEncodableInfo();
    _algorithmIdentifier = TextualInfo.GetInstance(signatureAlgorithm.Algorithm);
  }

  public Asn1Encodable ToPrimitive() => new List<IEncodableInfo> { _algorithmIdentifier, _additionalValue }
    .ToPrimitiveList().ToDerSequence()
    .ToAsn1Object();
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/AlgorithmInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toDLSequence
import com.jetbrains.signatureverifier.serialization.toEncodableInfo
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.ASN1ObjectIdentifier
import org.bouncycastle.asn1.ASN1Primitive
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.x509.AlgorithmIdentifier
import org.bouncycastle.operator.DefaultAlgorithmNameFinder

@Serializable
data class AlgorithmInfo(
  val name: String,
  val additionalValue: EncodableInfo? = null,
  val algorithmIdentifier: TextualInfo
) : EncodableInfo {
  constructor(signatureAlgorithm: AlgorithmIdentifier) : this(
    DefaultAlgorithmNameFinder().getAlgorithmName(
      signatureAlgorithm.algorithm as ASN1ObjectIdentifier
    ),
    signatureAlgorithm.parameters?.toASN1Primitive()?.toEncodableInfo(),
    TextualInfo.getInstance(signatureAlgorithm.algorithm)
  )

  private fun toDLSequence(): DLSequence {
    val list = mutableListOf(algorithmIdentifier.toPrimitive())
    if (additionalValue != null) {
      list.add(additionalValue.toPrimitive())
    }
    return list.toDLSequence()
  }

  override fun toPrimitive(): ASN1Primitive = toDLSequence().toASN1Primitive()
}